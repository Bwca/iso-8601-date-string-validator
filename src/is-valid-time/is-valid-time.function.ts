import { isValidZoneOffset } from '../is-valid-timezone-offset/is-valid-timezone-offset.function';
import { getStringSeparator } from '../utils/get-string-separator.function';

/**
 *
 * @param timeWithOffset hh:mm:ss(optional)Â±hh:mm(optional)
 * @param s separator between hours, minutes and optional seconds
 * @param isTimezoneCheckOn boolean flag, pass true not to check for valid timezone
 */
export function isValidTime(timeWithOffset: string, s = ':', isTimezoneCheckOn = false): boolean {

  const validator = new RegExp(
    `^([0-1]|2(?=([0-3])|4${s}00))[0-9]${s}[0-5][0-9](${s}([0-5]|6(?=0))[0-9])?(\.[0-9]{1,9})?$`,
  );

  if (!isTimezoneCheckOn || !/[Z+\-]/.test(timeWithOffset)) {
    return validator.test(timeWithOffset);
  }

  /** Case we got time in Zulu tz */
  if (/Z$/.test(timeWithOffset)) {
    return validator.test(timeWithOffset.replace('Z', ''));
  }

  const isPositiveTimezoneOffset = timeWithOffset.includes('+');
  const [time, offset] = timeWithOffset.split(/[+-]/);

  return validator.test(time) && isValidZoneOffset(offset, isPositiveTimezoneOffset, getStringSeparator(offset));

}
