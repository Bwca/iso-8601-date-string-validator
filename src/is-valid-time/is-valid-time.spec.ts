import { pad } from '../tests/utility-functions/pad.function';

import { isValidTime } from './is-valid-time.function';

/** max number of digits in hour/minute */
const maxDigits = 2;

test(`isValidTime. expect all hours with 0 minutes from 00:00 to 24:00 to validate true, 25+ hours to validate false`, () => {
  // Arrange
  const negativeResults: boolean[] = [];
  const positiveResults: boolean[] = [];

  // Act
  for (let hour = 0; hour <= 99; hour++) {
    const validationResult = isValidTime(
      `${pad(hour.toString(), maxDigits)}:00:00`,
    );

    hour < 25
      ? positiveResults.push(validationResult)
      : negativeResults.push(validationResult);
  }

  // Assert
  expect(positiveResults).not.toContain(false);
  expect(negativeResults).not.toContain(true);
  expect(positiveResults).toMatchSnapshot();
  expect(negativeResults).toMatchSnapshot();
});

test(`isValidTime. expect all minutes to validate beween 0 and 59, and to fail afterwards (from 60 to 99)`, () => {
  // Arrange
  const negativeResults: boolean[] = [];
  const positiveResults: boolean[] = [];

  // Act
  for (let minute = 0; minute <= 99; minute++) {
    const validationResult = isValidTime(
      `00:${pad(minute.toString(), maxDigits)}:00`,
    );
    minute < 60
      ? positiveResults.push(validationResult)
      : negativeResults.push(validationResult);
  }

  // Assert
  expect(positiveResults).not.toContain(false);
  expect(negativeResults).not.toContain(true);
  expect(positiveResults).toMatchSnapshot();
  expect(negativeResults).toMatchSnapshot();
});

test(`isValidTime. expect 24:00 to validate and 24:01-59 to fail`, () => {
  // Arrange
  const negativeResults: boolean[] = [];
  const positiveResults: boolean[] = [];

  // Act
  for (let minute = 0; minute <= 59; minute++) {
    const validationResult = isValidTime(
      `24:${pad(minute.toString(), maxDigits)}`,
    );
    minute === 0
      ? positiveResults.push(validationResult)
      : negativeResults.push(validationResult);
  }

  // Assert
  expect(positiveResults).not.toContain(false);
  expect(negativeResults).not.toContain(true);
  expect(positiveResults).toMatchSnapshot();
  expect(negativeResults).toMatchSnapshot();
});

test(`isValidTime. HH:mm to validate if empty separator is passed`, () => {
  // Arrange
  const time = '2200';
  const separator = '';
  let result: boolean;

  // Act
  result = isValidTime(time, separator);

  // Assert
  expect(result).toBe(true);
});

test(`isValidSeconds. expect seconds to validate only from 0 to 60, to fail when > 60`, () => {
  // Arrange
  const negativeResults: boolean[] = [];
  const positiveResults: boolean[] = [];

  // Act
  for (let second = 0; second < 100; second++) {
    const validationResult = isValidTime(
      `00:00:${pad(second.toString(), maxDigits)}.000`,
    );
    second <= 60
      ? positiveResults.push(validationResult)
      : negativeResults.push(validationResult);
  }

  // Assert
  expect(positiveResults).not.toContain(false);
  expect(negativeResults).not.toContain(true);
  expect(positiveResults).toMatchSnapshot();
  expect(negativeResults).toMatchSnapshot();
});

test(`isValidTime. expect time to validate when no separator provided`, () => {
  // Arrange
  const time = '00:00:00.111';
  const separator = ':';
  let result: boolean;

  // Act
  result = isValidTime(time, separator);

  // Assert
  expect(result).toBe(true);
});
