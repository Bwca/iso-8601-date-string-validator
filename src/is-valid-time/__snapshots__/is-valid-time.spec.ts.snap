// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isValidTime. expect 24:00 to validate and 24:01-59 to fail 1`] = `
Array [
  true,
]
`;

exports[`isValidTime. expect 24:00 to validate and 24:01-59 to fail 2`] = `
Array [
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
]
`;

exports[`isValidTime. expect all hours with 0 minutes from 00:00 to 24:00 to validate true, 25+ hours to validate false 1`] = `
Array [
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
]
`;

exports[`isValidTime. expect all hours with 0 minutes from 00:00 to 24:00 to validate true, 25+ hours to validate false 2`] = `
Array [
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
]
`;

exports[`isValidTime. expect all minutes to validate beween 0 and 59, and to fail afterwards (from 60 to 99) 1`] = `
Array [
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
]
`;

exports[`isValidTime. expect all minutes to validate beween 0 and 59, and to fail afterwards (from 60 to 99) 2`] = `
Array [
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
]
`;
